spec/policies/transaction_policy_spec.rb

```diff
@@ -1,75 +1,186 @@
+        it { is_expected.to forbid_action(:edit) }
+        it { is_expected.to forbid_action(:update) }
+        it { is_expected.to forbid_action(:destroy) }
+      end
+
+      context "and the activity does belong to the users organisation" do
+        before do
+          activity.update(organisation: user.organisation)
+        end
+
+        context "when there is no editable report" do
+          let(:report) { create(:report, state: :inactive) }
+
+          it { is_expected.to permit_action(:show) }
+
+          it { is_expected.to forbid_action(:create) }
+          it { is_expected.to forbid_action(:edit) }
+          it { is_expected.to forbid_action(:update) }
+          it { is_expected.to forbid_action(:destroy) }
+        end
+
+        context "when there is an editiable report" do
+          let(:report) { create(:report, state: :active) }
+
+          context "and the report is not for the organisation or fund of the activity" do
+            it { is_expected.to permit_action(:show) }
+
+            it { is_expected.to forbid_action(:create) }
+            it { is_expected.to forbid_action(:edit) }
+            it { is_expected.to forbid_action(:update) }
+            it { is_expected.to forbid_action(:destroy) }
+          end
+
+          context "and the report is for the organisation but not the fund of the activity" do
+            before do
+              report.update(organisation: activity.organisation)
+            end
+
+            it { is_expected.to permit_action(:show) }
+
+            it { is_expected.to forbid_action(:create) }
+            it { is_expected.to forbid_action(:edit) }
+            it { is_expected.to forbid_action(:update) }
+            it { is_expected.to forbid_action(:destroy) }
+          end
+
+          context "and the report is for the organisation and fund of the activity" do
+            before do
+              report.update(organisation: activity.organisation, fund: activity.associated_fund)
+            end
+
+            context "when the report is not the one in which the transaction was created" do
+              it { is_expected.to permit_action(:show) }
+              it { is_expected.to permit_action(:create) }
+
+              it { is_expected.to forbid_action(:edit) }
+              it { is_expected.to forbid_action(:update) }
+              it { is_expected.to forbid_action(:destroy) }
+            end
+
+            context "when the report is the one in which the transaction was created" do
+              before do
+                transaction.update(report: report)
+              end
+
+              it { is_expected.to permit_action(:show) }
+              it { is_expected.to permit_action(:create) }
+              it { is_expected.to permit_action(:edit) }
+              it { is_expected.to permit_action(:update) }
+
+              it { is_expected.to forbid_action(:destroy) }
+            end
+          end
+        end
+      end
+    end
```

Comment:
Amazing coverage of all possible scenarios on this spec!! Very clear and easy to follow

Reply:
👍 were you able to get a sense of this documenting the behaviour? I usually avoid this amount of nested contexts but I feel like this case actually warrants the complexity because there are so many factors.

Reply:
I was. It's true that there's a fair amount of nesting but everything was so well organised that it was easy to follow. And as you said, in this case it needed to be done this way. I wouldn't change anything.